### ISSUE TITLE
ARM64 platform compilation

### ISSUE THREAD

**Issue by mzky** [rescuezilla/rescuezilla#226]
I tried compiling under Debian 10.3 ARM64. Error message: " Unmet build dependencies: devscripts"
I tried installing dev-scripts using NPM and executed "aptitude install lxc". The same problem still arises.
Have you ever encountered such a situation?
I don't know if the ARM64 platform is the reason.

![image](https://user-images.githubusercontent.com/13345233/120949318-d182a300-c776-11eb-94e8-81cf256cdc32.

cpu:
root@super-PC:~# cat /proc/cpuinfo
processor       : 0
model name      : HUAWEI Kunpeng 920 2249K
BogoMIPS        : 200.00
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt lpae evtstrm
CPU implementer : 0x48
CPU architecture: 8
CPU variant     : 0x1
CPU part        : 0xd01
CPU revision    : 0

@shasheene I'm sorry! I created a new GitHub problem

**Comment by shasheene** [comment]
It's much more clear as a separate GitHub issue, thanks!

The `lxc` part is not required and was simply part of OAM7575's suggestions from #161 which are still being evaluated. The [official build instructions](https://github.com/rescuezilla/rescuezilla/blob/master/docs/build_instructions/BUILD.ISO.IMAGE.md) you referenced should largely be correct as of writing.

For Rescuezilla v2.2, I haven't tried any CPU architectures but AMD64, but the deb package only contains cross-platform assets like Python script files. In fact, the [standalone deb package install instructions](https://github.com/rescuezilla/rescuezilla/wiki/Installing-Rescuezilla-from-the-standalone-deb-package) should install and run fine on ARM64 (but I haven't tried this).

If you want to build a deb file (or even an the ISO image) from source, what does running `apt-get install devscripts` return? I suspect you don't have all the dependencies installed as per the build instructions.

**Comment by mzky** [comment]
> It's much more clear as a separate GitHub issue, thanks!
> 
> The `lxc` part is not required and was simply part of OAM7575's suggestions from #161 which are still being evaluated. The [official build instructions](https://github.com/rescuezilla/rescuezilla/blob/master/docs/build_instructions/BUILD.ISO.IMAGE.md) you referenced should largely be correct as of writing.
> 
> For Rescuezilla v2.2, I haven't tried any CPU architectures but AMD64, but the deb package only contains cross-platform assets like Python script files. In fact, the [standalone deb package install instructions](https://github.com/rescuezilla/rescuezilla/wiki/Installing-Rescuezilla-from-the-standalone-deb-package) should install and run fine on ARM64 (but I haven't tried this).
> 
> If you want to build a deb file (or even an the ISO image) from source, what does running `apt-get install devscripts` return? I suspect you don't have all the dependencies installed as per the build instructions.

You're right.I re-executed "apt-get install devscripts" to fix the problem
The Deb package compiled successfully.The above question is my misunderstanding. 

I am trying to compile ISO and for this I have also installed the following packages:
apt-get install autoconf automake libtool libattr1-dev
apt-get install build-essential debian-archive-keyring binfmt-support 
apt install libtool-bin libglobus-common-dev gawk

Still trying, hope I can succeed, and give other users a reference.
That's the problem right now, and I'm working on it.
![image](https://user-images.githubusercontent.com/13345233/120952956-a8661080-c77e-11eb-8fe9-1f4d52ef6d9d.png)

**Comment by shasheene** [comment]
Oh, Rescuezilla compiles `util-linux` and `partclone` to provide better backwards compatibility for very old images created by eg, Redo Backup and Recovery v0.9.2-v1.0.4. But you don't need that feature at all.

You can safely remove the dependency on `sfdisk.v2.20.1.amd64` and `partclone.restore.v0.2.43.amd64` from line 19 and 28 of the Makefile. Then it should build.

**Comment by shasheene** [comment]
I would also try to install _all_ that packages listed in the build instructions before proceeding:
```
sudo apt-get install git-lfs git make sudo \
                     rsync debootstrap gettext squashfs-tools dosfstools mtools xorriso \
                     memtest86+ devscripts debhelper checkinstall cmake \
                     grub-efi-amd64-bin grub-efi-ia32-bin grub-pc-bin \
                     shim-signed grub-efi-amd64-signed \
                     libtool-bin gawk pkg-config comerr-dev docbook-xsl e2fslibs-dev fuse \
                     libaal-dev libblkid-dev libbsd-dev libext2fs-dev libncurses5-dev \
                     libncursesw5-dev libntfs-3g88 libreadline-gplv2-dev libreadline5 \
                     libreiser4-dev libtinfo-dev libxslt1.1 nilfs-tools ntfs-3g ntfs-3g-dev \
                     quilt sgml-base uuid-dev vmfs-tools xfslibs-dev xfsprogs xml-core \
                     xsltproc
```

Hopefully all the packages are available on ARM64 Debian. By the way, any build you make on your Debian environment will not have Secure Boot work because the Debian GRUB package doesn't trust Ubuntu's Canonical-signed Linux kernel by default (as documented in "EFI Secure Boot on Debian and non-Canonical package environments") on the build instructions.

**Comment by shasheene** [comment]
I also recommend you build the program using `make clean && make hirsute`. This deletes the build folder (but not the downloaded deb package cache), and ensures a clean environment every rebuild.

**Comment by mzky** [comment]
The installation of RescueZilla's DEB package requires dependency:
apt install smbclient fsarchiver pv pigz nmap cifs-utils sshfs sshpass nfs-common yad pbzip2 pixz plzip lrzip zstd fatresize archivemount nbd-server nbd-client python3-whichcraft python3-babel clonezilla smartmontools nbdkit drbl samba-common libwbclient0  python3-tz  python-babel-localedata  liblz1 liblinear3  nmap-common  keyutils libtirpc3 libnfsidmap2 rpcbind libtirpc-common libblas3 libgfortran5 samba-libs ipcalc wakeonlan libsmbclient

There is no checkinstall installation package available under arm64.Try to download the official checkinstall compilation failed（https://asic-linux.com.mx/~izto/checkinstall/download.php）
Change checkinstall to make && make install（Makefile）, you can skip errors：
![image](https://user-images.githubusercontent.com/13345233/120981400-37d4e900-c7aa-11eb-9c90-7098d503a7a0.png)

Modify the debootstrap source code to meet the current system（build.sh）:
![image](https://user-images.githubusercontent.com/13345233/120982288-1d4f3f80-c7ab-11eb-9976-0a6a0b91a4fb.png)

make clean && make hirsute 
Stuck here：
![image](https://user-images.githubusercontent.com/13345233/120982948-d1e96100-c7ab-11eb-9c7b-a4fbb05e8175.png)

**Comment by shasheene** [comment]
> The installation of RescueZilla's DEB package requires dependency

That's correct. The standalone deb package requires a lot of dependencies. The [Github Wiki instructions for installing the standalone deb package](https://github.com/rescuezilla/rescuezilla/wiki/Installing-Rescuezilla-from-the-standalone-deb-package) suggest using `gdebi` to install the package, which will automatically download the required dependencies.

> Change checkinstall to make && make install（Makefile）, you can skip errors

You're [right that `checkinstall` is not available on Debian 10 Buster](https://packages.debian.org/search?suite=all&searchon=names&keywords=checkinstall). It's available on buster-backports and on other versions of Debian for some reason. But switching `checkinstall` to `make install` is acceptable here, so there's no need to install `checkinstall`. It just means it won't be managed by the system's package manager. Which is fine for your purposes.

> Modify the debootstrap source code to meet the current system（build.sh）:

It doesn't particularly matter that your machine is Debian 10 (Buster) for this step, you're trying to create an environment based on Ubuntu 21.04 (Hirsute) whatever your existing operating system.

This means your change to a geographically closer APT mirror **must** point to a *Ubuntu* APT mirror, not a Debian one. Also the modification to "buster" is **not** correct. It should still be `$CODENAME` as originally.

The `Makefile` is what sets `$CODENAME` and runs `build.sh`, where the code for "debootstrap part 1/2" is basically saying "please download me all the packages that are needed for `$CODENAME` and unpack the files, but _don't_ run anything yet".

> make clean && make hirsute
> Stuck here

This is too be expected given the previous step. Basically if you fix the "debootstrap step 1/2" part as explained above, then it should work.

**Comment by mzky** [comment]
Heartfelt thanks！
About "debootstrap part 1/2", I think the idea that it needs to be consistent with the current system is wrong.
I'll try again.

**Comment by mzky** [comment]
regrettably! It is always impossible to succeed in a Debian environment.

Try using Ubuntu.
Because Ubuntu officially does not have the latest ARM64 release. The following is a reference image:
https://mirrors.huaweicloud.com/ubuntukylin/ubuntukylin-focal-beta-desktop+arm64.iso
http://cdimage.ubuntu.com/ubuntu/daily-live/current/impish-desktop-arm64.iso 
http://cdimage.ubuntu.com/ubuntukylin/daily-live/current/impish-desktop-arm64.iso

The following is the closest approach to success:

```
sudo apt-get install  git make sudo \
                     rsync debootstrap gettext squashfs-tools dosfstools mtools xorriso \
                      devscripts debhelper checkinstall cmake \
                     grub-efi-arm64-bin  \
                      grub-efi-arm64-signed \
                     libtool-bin gawk pkg-config comerr-dev docbook-xsl e2fslibs-dev fuse \
                     libaal-dev libblkid-dev libbsd-dev libncurses5-dev \
                     libncursesw5-dev libreadline-gplv2-dev libreadline5 \
                     libreiser4-dev libtinfo-dev libxslt1.1 nilfs-tools ntfs-3g ntfs-3g-dev \
                     quilt sgml-base uuid-dev vmfs-tools xfslibs-dev xfsprogs xml-core \
                     xsltproc shim-signed  git-lfs
					 
sudo apt install ntfs-3g grub-common grub2-common e2fslibs memtester build-essential 
```
According to APT prompt:
libntfs-3g88 grub-efi-ia32-bin  shim-signed libext2fs-dev git-lfs memtest86+
libntfs-3g861 ==>ntfs-3g
grub-pc-bin grub-efi-ia32-bin ==>grub-common grub2-common
libext2fs-dev ==>e2fslibs
memtest86+ ==>memtester

Some variants do not have git-lfs. Can be installed using the official package:
https://github.com/git-lfs/git-lfs/releases/download/v2.13.3/git-lfs-linux-arm64-v2.13.3.tar.gz 

The Makefile to modify:
remove the dependency on sfdisk.v2.20.1.amd64 and partclone.restore.v0.2.43.amd64 from line 19 and 28 of the Makefile.
![image](https://user-images.githubusercontent.com/13345233/122167738-f3360580-cead-11eb-8696-bc56202be87d.png)
![image](https://user-images.githubusercontent.com/13345233/122167782-0052f480-ceae-11eb-9937-fb48f5cc417b.png)

The last problem is the same as with Debian:
![image](https://user-images.githubusercontent.com/13345233/122168110-70fa1100-ceae-11eb-9d49-b42258613d62.png)


The guess is that Ubuntu doesn't have a "debootstrap" package for Hirsute/Focal version.

**Comment by mzky** [comment]
Execute： 
```
debootstrap --arch=arm64 --foreign hirsute chroot http://mirrors.tuna.tsinghua.edu.cn/ubuntu-ports/
```
Can solve the above problems

But.
New problems keep cropping up!
Still trying...

**Comment by shasheene** [comment]
Yeah, creating a new build usually has a bunch of issues that need to be worked through. 

For Rescuezilla v2.3 I'm happy to put the work in to make an official ARM64 ISO image (potentially based on Debian 11 Bullseye, but maybe Ubuntu). But that will be for the November 2021 release, so I'm not intending on creating and debugging an ARM64 build before then.

If you do, then I'm happy to integrate the changes. But if it's too difficult, I will be having a look closer to November 2021.

**Comment by mzky** [comment]
Almost successful
![image](https://user-images.githubusercontent.com/13345233/123594110-e3a9ab80-d821-11eb-9312-a3ecccacef61.png)

**Comment by shasheene** [comment]
It says it's trying to find an i386-pc boot image (and other boot image files that may not exist for that platform)

**Comment by nnhq** [comment]
https://github.com/nnhq/rescuezilla/
build arm64 iso passing , run iso file on a pc seems not fully ok  ,i don't know what's going wrong.need all of you guys to fork it and fix it,don't pull request to this fork,feel free

**Comment by nnhq** [comment]
@mzky [github online workflow auto build passing，fork this ](https://github.com/nnhq/rescuezilla/) and run your workflow action，最终你会得到一个ISO文件，下载它。

**Comment by huapox** [comment]
![Image](https://github.com/user-attachments/assets/21de1867-08a7-48be-9e09-bfd0c090581f)

- fork构建生成 得ISO， https://github.com/gfk-sysenv/fk-nnhq-rescuezilla-arm64/releases/download/v25.06/rescuezilla.arm64.oracular.iso
-  ARM64-kunpeng920环境下跑KVM再启该ISO： 如上图lightdm, rc-local, nbc-server服务启动失败， 另请教 默认用户帐号密码是啥呢 @nnhq @shasheene

**Comment by shasheene** [comment]
@mzky, @nnhq, @huapox  just letting you know I've merged initial changes for the ARM64 build here: https://github.com/rescuezilla/rescuezilla/pull/642

It's bootable to a terminal interface, and will need more work to get `lightdm` (graphical stuff) working, improved GRUB language selection (currently text-only), and then actual Rescuezilla behavior.
